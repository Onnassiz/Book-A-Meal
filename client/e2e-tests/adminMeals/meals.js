/* eslint func-names: 0, no-unused-expressions: 0 */
const path = require('path');

const imagePath = path.join(__dirname, '../../assets/images/3.jpg');

const meals = {
  'Should sign user into the application': (client) => {
    client
      .url('http://localhost:3002/#/')
      .waitForElementVisible('header')
      .moveToElement('#login', 10, 10)
      .click('#login')
      .waitForElementVisible('h3')
      .assert.containsText('h3', 'Sign In')
      .setValue('#signInEmail', 'onnassiz@gmail.com')
      .setValue('#signInPassword', 'password')
      .click('input[type=submit]')
      .waitForElementVisible('header')
      .assert.visible('#caterer_dropdown')
      .assert.visible('#user_dropdown');
  },

  'Should navigate to the admin meals page': (client) => {
    client
      .moveToElement('#caterer_dropdown', 10, 10)
      .pause(1000)
      .waitForElementVisible('#admin_meals')
      .assert.visible('#admin_meals')
      .click('#admin_meals')
      .waitForElementVisible('#addMeal')
      .pause(2000)
      .assert.visible('#addMeal');
  },

  'Should show new meal modal form': (client) => {
    client
      .moveToElement('#addMeal', 10, 10)
      .pause(1000)
      .assert.containsText('#addMeal', 'Add Meal')
      .click('#addMeal')
      .pause(1000)
      .waitForElementVisible('.box')
      .assert.visible('.box')
      .pause(2000);
  },

  'Should fill meal form and create a new meal': (client) => {
    client
      .pause(1000)
      .setValue('#name', 'Fried Rice')
      .setValue('#price', 2000)
      .setValue('#category', 'Dinner')
      .setValue('#description', 'This is a good meal. It has all you need. Try it out.')
      .moveToElement('input[type=submit]', 10, 10)
      .pause(2000)
      .click('input[type=submit]')
      .waitForElementVisible('div.Toastify__toast.Toastify__toast--default', 5000)
      .assert.visible('div.Toastify__toast.Toastify__toast--default');
  },

  'Should show a new meal modal form again': (client) => {
    client
      .moveToElement('#addMeal', 10, 10)
      .assert.containsText('#addMeal', 'Add Meal')
      .click('#addMeal')
      .waitForElementVisible('.box')
      .assert.visible('.box')
      .pause(1000);
  },

  'Should fill another meal form and create a new meal': (client) => {
    client
      .pause(1000)
      .setValue('#name', 'Chicken')
      .setValue('#price', 2000)
      .setValue('#category', 'Dinner')
      .setValue('#description', 'This is a good meal. It has all you need. Try it out.')
      .moveToElement('input[type=submit]', 10, 10)
      .pause(2000)
      .click('input[type=submit]')
      .waitForElementVisible('div.Toastify__toast.Toastify__toast--default', 5000)
      .assert.visible('div.Toastify__toast.Toastify__toast--default');
  },

  'Should show a new meal modal form for the third time': (client) => {
    client
      .moveToElement('#addMeal', 10, 10)
      .assert.containsText('#addMeal', 'Add Meal')
      .click('#addMeal')
      .waitForElementVisible('.box')
      .assert.visible('.box')
      .pause(1000);
  },

  'Should fill the third meal form and create a new meal': (client) => {
    client
      .pause(1000)
      .setValue('#name', 'Beans')
      .setValue('#price', 1000)
      .setValue('#category', 'Dinner')
      .setValue('#description', 'This is a good meal for me and you.')
      .moveToElement('input[type=submit]', 10, 10)
      .pause(2000)
      .click('input[type=submit]')
      .waitForElementVisible('div.Toastify__toast.Toastify__toast--default', 5000)
      .assert.visible('div.Toastify__toast.Toastify__toast--default');
  },

  'Should show an update meal modal': (client) => {
    client
      .waitForElementVisible('#updateModal:first-of-type')
      .assert.visible('#updateModal:first-of-type')
      .click('#updateModal:first-of-type')
      .assert.visible('.box')
      .pause(1000);
  },

  'Should update the third meal form and update a new meal': (client) => {
    client
      .pause(1000)
      .clearValue('#name')
      .setValue('#name', 'Rice')
      .clearValue('#price')
      .setValue('#price', 1200)
      .moveToElement('input[type=submit]', 10, 10)
      .pause(1000)
      .click('input[type=submit]')
      .waitForElementVisible('div.Toastify__toast.Toastify__toast--default', 5000)
      .assert.visible('div.Toastify__toast.Toastify__toast--default');
  },

  'Should show delete modal and delete meal on button click': (client) => {
    client
      .assert.visible('#deleteModal:first-of-type')
      .click('#deleteModal:first-of-type')
      .waitForElementVisible('button.swal-button.swal-button--confirm.swal-button--danger', 10000)
      .pause(1000)
      .click('button.swal-button.swal-button--confirm.swal-button--danger')
      .waitForElementVisible('button.swal-button.swal-button--confirm', 15000)
      .pause(1000)
      .click('button.swal-button.swal-button--confirm')
      .waitForElementNotVisible('button.swal-button.swal-button--confirm', 20000)
      .assert.hidden('button.swal-button.swal-button--confirm')
      .pause(1000);
  },

  'Should show an image upload modal modal': (client) => {
    client
      .assert.visible('#addPhotoModal:first-of-type')
      .click('#addPhotoModal:first-of-type')
      .waitForElementVisible('#dropzoneHeader')
      .assert.containsText('#dropzoneHeader', 'Drop Image Here')
      .pause(1000)
      .moveToElement('#upload_image', 10, 10)
      .pause(1000)
      .setValue('input[type=file]', imagePath)
      .waitForElementVisible('div.Toastify__toast.Toastify__toast--default', 15000)
      .assert.visible('div.Toastify__toast.Toastify__toast--default')
      .pause(1000);
    client.end();
  },
};

export default meals;
