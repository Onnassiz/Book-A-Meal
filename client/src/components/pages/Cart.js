import React, { Component } from 'react';
import CartContainer from '../container/CartContainer';
import NavBarContainer from '../container/NavBarContainer';

class Cart extends Component {
  render() {
    return (
			<div>
				<NavBarContainer page="cart" />
				<CartContainer />
			</div>
    );
  }
}

export default Cart;
