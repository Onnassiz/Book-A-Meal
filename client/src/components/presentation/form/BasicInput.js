import React from 'react';
import PropTypes from 'prop-types';

export const BasicInput = props => (
	<div className="inputBox">
		<label htmlFor={props.name}>{props.label}</label>
		<input autoComplete={props.name} className={props.hasError ? 'has-error' : ''} onChange={props.onChange} value={props.value} type={props.type} name={props.name} id={props.name} />
	</div>
);

export const TextArea = props => (
	<div className="inputBox">
		<label htmlFor={props.name}>{props.label}</label>
		<textarea className={props.hasError ? 'has-error' : ''} name={props.name} id={props.name} cols="30" rows="6" onChange={props.onChange} value={props.value} />
	</div>
);

export const Checkbox = props => (
	<label className="checkbox"> {props.meal.name} <span style={{ float: 'right', color: '#7a604a', fontSize: 10 }}>Added by {props.meal.caterer}</span>
		<input type="checkbox" value={props.meal.id} onChange={props.onChange} checked={props.isChecked} /><span className="check" />
	</label>
);

BasicInput.propTypes = {
	name: PropTypes.string.isRequired,
	label: PropTypes.string.isRequired,
	type: PropTypes.string.isRequired,
	value: PropTypes.string.isRequired,
	onChange: PropTypes.func.isRequired,
	hasError: PropTypes.bool,
};

BasicInput.defaultProps = {
	hasError: false,
};

TextArea.propTypes = {
	name: PropTypes.string.isRequired,
	label: PropTypes.string.isRequired,
	value: PropTypes.string.isRequired,
	onChange: PropTypes.func.isRequired,
	hasError: PropTypes.bool,
};

Checkbox.propTypes = {
	meal: PropTypes.object.isRequired,
	isChecked: PropTypes.bool.isRequired,
	onChange: PropTypes.func.isRequired,
};

TextArea.defaultProps = {
	hasError: false,
};

