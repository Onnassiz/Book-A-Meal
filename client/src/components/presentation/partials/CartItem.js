import React from 'react';
import PropTypes from 'prop-types';
import empty from 'is-empty';
import { numberWithCommas } from '../../../utilities/functions';


const CartItem = (props) => {
  let image = props.meal.imageUrl;
  image = image === null ? null : image.replace('upload/', 'upload/c_scale,w_100/');
  return (
		<div className="cart-item" key={props.meal.id}>
			<p><span className="name">{props.meal.name}</span></p>
			<div className="image">
				{empty(props.meal.imageUrl) ? <div /> : <img src={image} alt={props.meal.name} />}
			</div>
			<div className="details">
				<p>Price: &#8358;{numberWithCommas(props.meal.price)}</p>
				<p>Total Price: &#8358;{numberWithCommas(props.meal.totalPrice)}</p>
				<p>Quantity: <input type="number" min="1" value={props.meal.units} onChange={props.updateUnits} /></p>
				<p><button onClick={props.deleteFromCart}><i className="ion-trash-b" /> Remove</button></p>
			</div>
			<div style={{ clear: 'both', paddingTop: 1, width: 250 }}>
				<hr />
			</div>
		</div>
  );
};

CartItem.propTypes = {
  meal: PropTypes.object.isRequired,
  updateUnits: PropTypes.func.isRequired,
  deleteFromCart: PropTypes.func.isRequired,
};

export default CartItem;
