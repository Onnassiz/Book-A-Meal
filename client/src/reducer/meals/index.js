import { SET_MEAL, SET_MEAL_SERVER_ERRORS } from './actions';

const initialState = {
  meals: [],
};

export default (state = initialState, action) => {
  switch (action.type) {
    case SET_MEAL:
      return { ...state, meals: action.meals };
    case SET_MEAL_SERVER_ERRORS:
      return { ...state, errors: action.errors };
    default:
      return state;
  }
};
