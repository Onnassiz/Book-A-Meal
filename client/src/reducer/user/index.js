import { SET_USER, SET_SIGN_UP_SERVER_ERRORS, SET_SIGN_IN_SERVER_ERRORS } from './action';

const initialState = {};

export default (state = initialState, action) => {
	switch (action.type) {
	case SET_USER:
		return { name: action.name, email: action.email, role: action.role };
	case SET_SIGN_UP_SERVER_ERRORS:
		return { signUpErrors: action.errors };
	case SET_SIGN_IN_SERVER_ERRORS:
		return { signInErrors: action.errors };
	default:
		return state;
	}
};
