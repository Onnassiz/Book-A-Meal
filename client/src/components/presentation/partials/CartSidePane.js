import React from 'react';
import PropTypes from 'prop-types';
import empty from 'is-empty';
import SlidingPane from 'react-sliding-pane';
import CartItem from './../partials/CartItem';

const CartSidePane = props => (
  <SlidingPane
    isOpen={props.isOpen}
    title="Cart Info"
    from="left"
    width="320px"
    onRequestClose={props.closeCartPane}
  >
    <div>
      {props.cart.updateMode ?
        <div id="empty-cart">
          <h2>Modify Order</h2>
        </div> : ''}
      <div>
        {empty(props.cart.cart) ?
          <div id="empty-cart">
            <h2>Your cart is empty</h2>
          </div> : props.cart.cart.map(meal =>
            (<CartItem
              deleteFromCart={() => props.removeFromCart(meal.id)}
              meal={meal}
              key={meal.id + meal.caterer}
              updateUnits={e => props.updateUnits(e, meal)}
            />))}
      </div>
      <div className="checkout">
        {empty(props.cart.cart) ? '' :
        <div>
          <div className="cart_summary_body">
            <p className="num">Number of Items: {props.cart.length}</p>
            <p className="num">Delivery Fee: &#8358;0.00</p>
            <p className="num">Discount: &#8358;0.00</p>
            <br />
            <br />
            <p className="num">Total Amount: <b>&#8358;{props.cart.totalPrice}.00</b></p>
            <br />
            <p>
              <button onClick={props.emptyCart}>Empty Cart</button>
              <button onClick={props.checkout}>{props.cart.updateMode ? 'Next' : 'Checkout'}</button>
            </p>
          </div>
        </div>}
      </div>
    </div>
  </SlidingPane>
);

CartSidePane.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  closeCartPane: PropTypes.func.isRequired,
  cart: PropTypes.object.isRequired,
  totalPrice: PropTypes.string.isRequired,
  emptyCart: PropTypes.func.isRequired,
  checkout: PropTypes.func.isRequired,
};

export default CartSidePane;
