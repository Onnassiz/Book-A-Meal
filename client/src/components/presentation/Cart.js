import React, { Component } from 'react';
import PropTypes from 'prop-types';
import empty from 'is-empty';
import Modal from 'react-modal';
import { toast } from 'react-toastify';
import { addProfileImageModalView } from './../../utilities/modalStyles';
import { BasicInput, TextArea } from './form/BasicInput';
import SubmitButton from '../presentation/form/SubmitButton';
import CartSidePane from './partials/CartSidePane';
import { numberWithCommas } from '../../utilities/functions';
import validateOrder from '../../utilities/validateOrder';


class Cart extends Component {
  constructor(props) {
    super(props);
    this.state = {
      id: '',
      totalPrice: 0,
      isPaneOpenLeftCart: false,
      showCheckoutModal: false,
      telephone: '',
      address: '',
      errors: {},
    };
    this.onChange = this.onChange.bind(this);
    this.showCart = this.showCart.bind(this);
    this.removeFromCart = this.removeFromCart.bind(this);
    this.updateUnits = this.updateUnits.bind(this);
    this.showCheckoutForm = this.showCheckoutForm.bind(this);
    this.closeCheckoutForm = this.closeCheckoutForm.bind(this);
    this.placeOrder = this.placeOrder.bind(this);
    this.afterOpenModal = this.afterOpenModal.bind(this);
  }

  onChange(e) {
    this.setState({ [e.target.name]: e.target.value });
  }

  removeFromCart(mealId) {
    const { deleteFromCart, user } = this.props;
    deleteFromCart(mealId, user.id);
  }

  updateUnits(e, currentMeal) {
    const { value } = e.target;
    if (value > 0) {
      const { updateCart, user } = this.props;
      const meal = currentMeal;
      meal.units = parseInt(value, 10);
      meal.totalPrice = value * meal.price;
      updateCart(meal, user.id);
    }
  }

  showCart(state) {
    this.props.setCartState(state);
  }

  closeCheckoutForm() {
    this.setState({
      showCheckoutModal: false,
    });
  }

  showCheckoutForm() {
    this.setState({
      showCheckoutModal: true,
      isPaneOpenLeftCart: false,
    });
  }

  isValid() {
    this.setState({ errors: {} });
    const { errors, isValid } = validateOrder(this.state);
    if (!isValid) {
      this.setState({ errors });
    }
    return isValid;
  }

  afterOpenModal() {
    const { cart } = this.props;
    this.showCart(false);
    if (cart.updateMode) {
      this.setState({
        id: cart.orderId,
        address: cart.address,
        telephone: cart.contact,
      });
    }
  }

  createOrder(order) {
    const { postOrder } = this.props;
    postOrder(order).then((response) => {
      const { emptyCart } = this.props;
      if (response.status === 201) {
        this.resetFields();
        this.closeCheckoutForm();
        emptyCart();
        toast('Your order has been placed. Expect delivery shortly.');
      }
    });
  }

  updateOrder(order) {
    const { updateOrder } = this.props;
    updateOrder(order).then((response) => {
      const { emptyCart } = this.props;
      if (response.status === 200) {
        this.resetFields();
        this.closeCheckoutForm();
        emptyCart();
        toast('Your order has been updated.');
      }
    });
  }

  placeOrder(e) {
    e.preventDefault();
    if (this.isValid()) {
      const { cart } = this.props;
      const order = this.processPayload();
      if (cart.updateMode) {
        this.updateOrder(order);
      } else {
        this.createOrder(order);
      }
    }
  }

  processPayload() {
    const { cart } = this.props;
    const meals = [];
    cart.cart.forEach((meal) => {
      meals.push({
        mealId: meal.id,
        price: meal.price,
        units: meal.units,
      });
    });
    const payload = {
      id: cart.orderId,
      contact: this.state.telephone,
      address: this.state.address,
      meals,
    };
    return payload;
  }

  resetFields() {
    this.setState({ telephone: '', address: '' });
  }

  renderCart() {
    const { cart } = this.props;
    return (
      <CartSidePane
        isOpen={cart.isOpen}
        closeCartPane={() => this.showCart(false)}
        cart={cart}
        updateUnits={this.updateUnits}
        checkout={this.showCheckoutForm}
        removeFromCart={this.removeFromCart}
        totalPrice={numberWithCommas(this.state.totalPrice)}
        emptyCart={this.props.emptyCart}
      />
    );
  }

  renderFormBody() {
    const { orders, formState } = this.props;
    return (
      <div>
        <div className="col-12">
          <a onClick={this.closeCheckoutForm} style={{ float: 'right' }}><i style={{ fontSize: 25 }} className="material-icons">close</i></a>
        </div>
        <div className="box">
          <div className="show-errors">
            <ul>
              {Object.keys(this.state.errors)
                .map(item => <li key={item}>{this.state.errors[item]}</li>)}
              {empty(orders.errors) ? '' :
                Object.keys(orders.errors).map(item => <li key={item}>{orders.errors[item]}</li>)}
            </ul>
          </div>
          <form onSubmit={this.placeOrder}>
            <BasicInput type="text" label="Phone Number" name="telephone" value={this.state.telephone} onChange={this.onChange} hasError={this.state.errors.telephone !== undefined} />
            <TextArea name="address" type="text" label="Address" value={this.state.address} onChange={this.onChange} hasError={this.state.errors.address !== undefined} />
            <SubmitButton value={this.state.updateMode ? 'Update' : 'Submit'} isLoading={formState.isLoading} />
          </form>
        </div>
      </div>
    );
  }

  renderCheckoutForm() {
    return (
      <Modal
        isOpen={this.state.showCheckoutModal}
        closeTimeoutMS={1}
        onAfterOpen={this.afterOpenModal}
        style={addProfileImageModalView}
        shouldCloseOnEsc
        shouldCloseOnOverlayClick
        onRequestClose={this.closeCheckoutForm}
        contentLabel="Modal"
      >
        {this.renderFormBody()}
      </Modal>
    );
  }

  renderFixedCart() {
    const { cart } = this.props;
    return (
      <div className={cart.cart.length ? 'dynamic-badge cart-control' : 'cart-control'} data-badge={cart.cart.length}>
        <i className="ion-ios-cart-outline" onClick={() => this.showCart(true)} />
      </div>
    );
  }

  render() {
    return (
      <div>
        {this.renderCheckoutForm()}
        {this.renderFixedCart()}
        {this.renderCart()}
      </div>
    );
  }
}

Cart.propTypes = {
  emptyCart: PropTypes.func.isRequired,
  cart: PropTypes.object.isRequired,
  user: PropTypes.object.isRequired,
  orders: PropTypes.object.isRequired,
  formState: PropTypes.object.isRequired,
  deleteFromCart: PropTypes.func.isRequired,
  updateOrder: PropTypes.func.isRequired,
  updateCart: PropTypes.func.isRequired,
  setCartState: PropTypes.func.isRequired,
  postOrder: PropTypes.func.isRequired,
};

export default Cart;
