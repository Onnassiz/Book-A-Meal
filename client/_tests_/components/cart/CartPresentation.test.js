import React from 'react';
import { shallow } from 'enzyme';
import SlidingPane from 'react-sliding-pane';
import toJson from 'enzyme-to-json';
import Modal from 'react-modal';
import Cart from '../../../src/components/presentation/Cart';
import user from '../../objectProps/user.props';
import { BasicInput } from '../../../src/components/presentation/form/BasicInput';
import CartSidePane from '../../../src/components/presentation/partials/CartSidePane';
import CartItem from '../../../src/components/presentation/partials/CartItem';
import { getMealsMock } from '../../objectProps/meals.props';

const cart = {
  cart: getMealsMock(1),
  owner: '',
  orderId: '',
  address: '',
  contact: '',
  isOpen: false,
  totalPrice: 0,
  updateMode: false,
};

const props = {
  orders: {},
  formState: { isLoading: false },
  user,
  cart,
  emptyCart: jest.fn(),
  deleteFromCart: jest.fn(),
  updateOrder: () => new Promise(resolve => resolve({ status: 200 })),
  updateCart: jest.fn(),
  setCartState: jest.fn(),
  postOrder: () => new Promise(resolve => resolve({ status: 201 })),
};

const wrapper = shallow(<Cart {...props} />);

describe('<Cart />', () => {
  it('Should render the Cart container successfully', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Should trigger onChange when input field is updated', () => {
    wrapper.find(BasicInput).dive().find('input').simulate('change', { target: { name: 'address', value: 'This is the address' } });
    expect(wrapper.state().address).toBe('This is the address');
  });

  it('Should trigger removeFromCart when remove button is clicked', () => {
    const removeFromCartSpy = jest.spyOn(wrapper.instance(), 'removeFromCart');

    wrapper.setState({});

    wrapper.find(CartSidePane).dive().find(CartItem).first()
      .dive()
      .find('.cart-button')
      .simulate('click');

    expect(removeFromCartSpy).toHaveBeenCalled();
  });

  it('Should trigger updateUnits when number input field button is clicked', () => {
    const updateUnitsSpy = jest.spyOn(wrapper.instance(), 'updateUnits');

    wrapper.setState({});

    wrapper.find(CartSidePane).dive().find(CartItem).first()
      .simulate('updateUnits');

    expect(updateUnitsSpy).toHaveBeenCalled();
  });

  it('Should trigger showCheckoutForm and showCheckoutForm when checkout button is click', () => {
    const showCheckoutFormSpy = jest.spyOn(wrapper.instance(), 'showCheckoutForm');

    wrapper.setState({});

    wrapper.find(CartSidePane).dive().find('.checkout').find('button')
      .last()
      .simulate('click');

    expect(wrapper.state().showCheckoutModal).toBe(true);
    expect(wrapper.state().isPaneOpenLeftCart).toBe(false);
    expect(showCheckoutFormSpy).toHaveBeenCalled();
  });

  it('Should trigger onRequestClose when CartSidePane close button is clicked', () => {
    const showCartSpy = jest.spyOn(wrapper.instance(), 'showCart');

    wrapper.setState({});

    wrapper.find(CartSidePane).dive().find(SlidingPane)
      .simulate('requestClose');

    expect(showCartSpy).toHaveBeenCalled();
  });

  it('Should trigger placeOrder and show errors when checkout form is submitted with empty fields', () => {
    const placeOrderSpy = jest.spyOn(wrapper.instance(), 'placeOrder');

    wrapper.setState({ telephone: '', address: '' });

    wrapper.find('form').simulate('submit', { preventDefault: jest.fn() });

    expect(wrapper.state().errors.telephone).toBe('The phone number field is required');
    expect(wrapper.state().errors.address).toBe('The address field is required');
    expect(placeOrderSpy).toHaveBeenCalled();
  });

  it('Should trigger placeOrder and createOrder when checkout button is clicked', (done) => {
    const placeOrderSpy = jest.spyOn(wrapper.instance(), 'placeOrder');
    const createOrderSpy = jest.spyOn(wrapper.instance(), 'createOrder');

    wrapper.setState({ telephone: '080232232232', address: 'This is an address' });

    wrapper.find('form').simulate('submit', { preventDefault: jest.fn() });

    setImmediate(() => {
      expect(wrapper.state().showCheckoutModal).toBe(false);
      expect(wrapper.state().address).toBe('');
      expect(wrapper.state().telephone).toBe('');
      expect(placeOrderSpy).toHaveBeenCalled();
      expect(createOrderSpy).toHaveBeenCalled();
      done();
    });
  });

  it('Should trigger placeOrder and updateOrder when checkout button is clicked', (done) => {
    const placeOrderSpy = jest.spyOn(wrapper.instance(), 'placeOrder');
    const updateOrderSpy = jest.spyOn(wrapper.instance(), 'updateOrder');

    wrapper.setState({ telephone: '080232232232', address: 'This is an address' });

    cart.updateMode = true;
    wrapper.setProps({ cart });

    wrapper.find('form').simulate('submit', { preventDefault: jest.fn() });

    setImmediate(() => {
      expect(wrapper.state().showCheckoutModal).toBe(false);
      expect(wrapper.state().address).toBe('');
      expect(wrapper.state().telephone).toBe('');
      expect(placeOrderSpy).toHaveBeenCalled();
      expect(updateOrderSpy).toHaveBeenCalled();
      done();
    });
  });

  it('Should trigger afterOpenModal when checkout modal shows', () => {
    const afterOpenModalSpy = jest.spyOn(wrapper.instance(), 'afterOpenModal');

    wrapper.setState({});

    wrapper.find(Modal).simulate('afterOpen');

    expect(afterOpenModalSpy).toHaveBeenCalled();
  });

  it('Should trigger renderFixedCart to show cart', () => {
    const showCartSpy = jest.spyOn(wrapper.instance(), 'showCart');

    wrapper.setState({});

    wrapper.find('.ion-ios-cart-outline').simulate('click');

    expect(showCartSpy).toHaveBeenCalled();
  });
});
