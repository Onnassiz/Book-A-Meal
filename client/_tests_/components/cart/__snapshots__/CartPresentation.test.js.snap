// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart /> Should render the Cart container successfully 1`] = `
<div>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    closeTimeoutMS={1}
    contentLabel="Modal"
    isOpen={false}
    onAfterOpen={[Function]}
    onRequestClose={[Function]}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
    style={
      Object {
        "content": Object {
          "WebkitOverflowScrolling": "touch",
          "background": "#ffffff",
          "border": "none",
          "borderRadius": "6px",
          "bottom": "30%",
          "left": "33%",
          "outline": "none",
          "overflow": "auto",
          "padding": "15px",
          "position": "absolute",
          "right": "33%",
          "top": "15%",
        },
        "overlay": Object {
          "backgroundColor": "rgba(0, 0, 0, 0.4)",
          "bottom": 0,
          "left": 0,
          "position": "fixed",
          "right": 0,
          "top": 0,
        },
      }
    }
  >
    <div>
      <div
        className="col-12"
      >
        <a
          onClick={[Function]}
          style={
            Object {
              "float": "right",
            }
          }
        >
          <i
            className="material-icons"
            style={
              Object {
                "fontSize": 25,
              }
            }
          >
            close
          </i>
        </a>
      </div>
      <div
        className="box"
      >
        <ShowErrors
          clientErrors={Object {}}
          serverErrors={Object {}}
        />
        <form
          onSubmit={[Function]}
        >
          <BasicInput
            hasError={false}
            label="Phone Number"
            min=""
            name="telephone"
            onChange={[Function]}
            type="text"
            value=""
          />
          <TextArea
            hasError={false}
            label="Address"
            name="address"
            onChange={[Function]}
            type="text"
            value=""
          />
          <SubmitButton
            isLoading={false}
            value="Submit"
          />
        </form>
      </div>
    </div>
  </Modal>
  <div
    className="dynamic-badge cart-control"
    data-badge={1}
  >
    <i
      className="ion-ios-cart-outline"
      onClick={[Function]}
    />
  </div>
  <CartSidePane
    cart={
      Object {
        "address": "",
        "cart": Array [
          Object {
            "category": "dolor sit",
            "caterer": "Ben",
            "description": "Maecenas ut massa quis augue luctus tincidunt. Nulla mollis molestie lorem. Quisque ut erat. Curabitur gravida nisi at nibh.",
            "id": "84c6d80c-3d4a-4772-856c-0b3825bc5333",
            "imageUrl": "http://thisImage.jpeg",
            "name": "Malvaceae",
            "price": 400,
            "totalPrice": 2000,
            "units": 2,
            "userId": "2794fddd-14f3-4ec9-a1dc-88ad2bc649da",
          },
        ],
        "contact": "",
        "isOpen": false,
        "orderId": "",
        "owner": "",
        "totalPrice": 0,
        "updateMode": false,
      }
    }
    checkout={[Function]}
    closeCartPane={[Function]}
    emptyCart={[MockFunction]}
    isOpen={false}
    removeFromCart={[Function]}
    totalPrice="0"
    updateUnits={[Function]}
  />
</div>
`;
